### PR TITLE
[interp] Add intrinsics for some common operations

### PR DESCRIPTION
**Pull request by BrzVlad**
These are mostly around string operations and are meant to be a quick boost in performance for some common blazor scenarios, around rendering and json parsing, before interp optimizations catch up. Even if the interpreter will have more advanced optimization passes, it is likely that it would still be much slower on these low level operations. Long term we would have methods like these AOT-ed, but, depending on how easy to maintain and how effective these turn out to be, we could also stick to intrinsifying some bcl code in the future.

### CODE REVIEW

File: src/mono/mono/mini/interp/intrinsics.c [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+	return digits;
+}
+
+static guint32
+interp_intrins_math_divrem (guint32 a, guint32 b, guint32 *result)
+{
+	guint32 div = a / b;
+	*result = a - (div * b);
+	return div;
+}
```
Comment (EgorBo) [inline src/mono/mono/mini/interp/intrinsics.c:102]: can be just
```c
*result = a % b;
return a / b;
```
C/C++ compilers are able to recognize a single div there https://godbolt.org/z/-7jsRb

File: src/mono/mono/mini/interp/transform.c [hunk 2]
```diff
@@ -1479,6 +1479,37 @@ interp_handle_intrinsics (TransformData *td, MonoMethod *target_method, MonoClas
 	} else if (in_corlib && !strcmp (klass_name_space, "System") && !strcmp (klass_name, "ByReference`1")) {
 		g_assert (!strcmp (tm, "get_Value"));
 		*op = MINT_INTRINS_BYREFERENCE_GET_VALUE;
+	} else if (in_corlib && !strcmp (klass_name_space, "System") && !strcmp (klass_name, "Marvin")) {
+		if (!strcmp (tm, "Block"))
+			*op = MINT_INTRINS_MARVIN_BLOCK;
```
Comment (EgorBo) [inline src/mono/mono/mini/interp/transform.c:1484]: I'd just intrinsify `String.GetHashCode` and implement it using the non-random algorithm like this: https://github.com/dotnet/runtime/blob/master/src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs#L781-L783 should be way faster and that type of attack usually back-end only

File: src/mono/mono/mini/interp/transform.c [hunk 3]
```diff
@@ -1479,6 +1479,37 @@ interp_handle_intrinsics (TransformData *td, MonoMethod *target_method, MonoClas
...
+	} else if (in_corlib && !strcmp (klass_name_space, "System.Text.Unicode") && !strcmp (klass_name, "Utf16Utility")) {
```
Comment (MichalStrehovsky) [inline src/mono/mono/mini/interp/transform.c:1485]: Is this going to silently regress perf if Utf16Utility is ever refactored and things in it shuffled/renamed? This is not a public API so the risk of that is pretty high.

Cc @GrabYourPitchforks

File: src/mono/mono/mini/interp/transform.c (hunk 2 above)
Comment (BrzVlad) [reply src/mono/mono/mini/interp/transform.c:1484]: Doesn't sound like a good idea to have the interpreter just use a different hash function compared to jit/aot or coreclr. Arguably, maybe I should intrinsify both `Marvin.ComputeHash32` and  `Marvin.ComputeHash32OrdinalIgnoreCase`.

File: src/mono/mono/mini/interp/transform.c (hunk 3 above)
Comment (BrzVlad) [reply src/mono/mono/mini/interp/transform.c:1485]: Unfortunately yes. We should spot potential regressions with the daily perf runs that we will do. Hopefully we will not make use of this technique on a large enough scale for it to become problematic.

File: src/mono/mono/mini/interp/transform.c (hunk 2 above)
Comment (EgorBo) [reply src/mono/mono/mini/interp/transform.c:1484]: I'd suggest to use this one for all Xamarin workloads and all modes including AOT.

also it doesn't matter if two different implementations return different hash codes - string.GetHashCode already returns different hash codes for the same input between sessions - it uses a random seed.

File: src/mono/mono/mini/interp/transform.c (hunk 3 above)
Comment (akoeplinger) [reply src/mono/mono/mini/interp/transform.c:1485]: can we mark the method with `[Intrinsic]` or add some comment to make the chance of this happening a bit less likely?

File: src/mono/mono/mini/interp/transform.c (hunk 2 above)
Comment (akoeplinger) [reply src/mono/mono/mini/interp/transform.c:1484]: @EgorBo I think we should discuss that in a separate issue.

File: src/libraries/System.Private.CoreLib/src/System/Marvin.cs [hunk 4]
```diff
@@ -215,6 +215,7 @@ public static int ComputeHash32(ref byte data, uint count, uint p0, uint p1)
             goto DoFinalRoundsAndReturn;
         }
 
+        [Intrinsic]
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
```
Comment (EgorBo) [inline src/libraries/System.Private.CoreLib/src/System/Marvin.cs:219]: in theory this attribute slows ryujit a bit for no reason, I'd just leave a comment or wrap it with #if mono

**Review by vargaz — APPROVED** [review]

**Pull request merged**